src/plugins/dashboard/public/dashboard_app/top_nav/add_panel_action_menu_items.ts

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+import type { ActionExecutionContext, Action } from '@kbn/ui-actions-plugin/public';
+import type { EmbeddableFactory } from '@kbn/embeddable-plugin/public';
+import { addPanelMenuTrigger } from '../../triggers';
+
+const onAddPanelActionClick =
+  (action: Action, context: ActionExecutionContext<object>, closePopover: () => void) =>
+  (event: React.MouseEvent) => {
+    closePopover();
+    if (event.currentTarget instanceof HTMLAnchorElement) {
+      if (
+        !event.defaultPrevented && // onClick prevented default
+        event.button === 0 &&
+        (!event.currentTarget.target || event.currentTarget.target === '_self') &&
+        !(event.metaKey || event.altKey || event.ctrlKey || event.shiftKey)
+      ) {
+        event.preventDefault();
+        action.execute(context);
+      }
+    } else action.execute(context);
+  };
+
+export const getAddPanelActionMenuItems = (
+  actions: Array<Action<object>> | undefined,
+  createNewEmbeddable: (embeddableFactory: EmbeddableFactory) => void,
+  deleteEmbeddable: (embeddableId: string) => void,
+  closePopover: () => void
+) => {
+  const actionsWithContext =
+    actions?.map((action) => ({
+      action,
+      context: {
+        createNewEmbeddable,
+        deleteEmbeddable,
+      },
+      trigger: addPanelMenuTrigger,
+    })) ?? [];
+
+  return actionsWithContext?.map((item) => {
```

Comment:
Is actionsWithContext needed or could you just return map of actions and do everything in a single loop? I do not understand why this logic was broken into 2 loops

Reply:
No need at all, another leftover from my previous implementation. Done https://github.com/elastic/kibana/pull/171973/commits/92cab39c055adeff6ec18244f140c83ca83d5c4c